### PR TITLE
[SuperTextLayout] Paint text selection with BorderRadius

### PR DESCRIPTION
**Pull request by domesticmouse**

### CODE REVIEW

**Comment by domesticmouse** [comment]
@matthew-carroll I'm curious what tests you would like for this?

**Comment by domesticmouse** [comment]
I think the goldens i generated on macOS aren't compatible with CI. As a googler, I can't use docker...

**Comment by matthew-carroll** [comment]
@domesticmouse you can't use Docker on your personal machine? Or you can't use Docker on a G3 machine and that's all you've got?

File: super_text_layout/lib/src/text_selection_layer.dart [hunk 1]
```diff
@@ -132,12 +133,14 @@ class TextSelectionPainter extends CustomPainter {
   TextSelectionPainter({
     required this.textLayout,
     required this.textSelection,
+    required this.borderRadius,
```
Comment (matthew-carroll) [inline super_text_layout/lib/src/text_selection_layer.dart:136]: This is going to be a breaking change, as-is. Please switch this to a default value of zero so that nothing upstream breaks by adding this.

**Comment by domesticmouse** [comment]
I don't currently have a personal dev machine. Can you update the goldens in this PR to make the CI happy?

File: super_text_layout/test_goldens/text_selection_layer_test.dart [hunk 2]
```diff
@@ -77,4 +77,76 @@ void main() {
...
+        tester,
+        beneathBuilder: (context, textLayout) {
+          return TextLayoutEmptyHighlight(
+            textLayout: textLayout,
+            style: selectionStyle,
+          );
+        },
+      );
+
+      await screenMatchesGolden(tester, "TextSelectionLayer_small-highlight-when-empty-border-radius");
```
Comment (matthew-carroll) [inline super_text_layout/test_goldens/text_selection_layer_test.dart:134]: Is this golden image correct for this test? I don't see rounded corners on the empty highlight.

File: super_text_layout/lib/src/text_selection_layer.dart (hunk 1 above)
Comment (domesticmouse) [reply super_text_layout/lib/src/text_selection_layer.dart]: Done.

File: super_text_layout/test_goldens/text_selection_layer_test.dart (hunk 2 above)
Comment (domesticmouse) [reply super_text_layout/test_goldens/text_selection_layer_test.dart:134]: The empty case was a separate codepath. Found and updated.

**Comment by matthew-carroll** [comment]
@angelosilvestre can you try generating the goldens for this PR?

File: super_text_layout/lib/src/text_selection_layer.dart [hunk 3]
```diff
@@ -156,11 +165,18 @@ class TextSelectionPainter extends CustomPainter {
...
+          bottomRight: borderRadius.bottomRight);
 
-      canvas.drawRect(
+      canvas.drawRRect(
         // Note: If the rect has no width then we've selected an empty line. Give
         //       that line a slight width for visibility.
-        rect.width > 0 ? rect : Rect.fromLTWH(rect.left, rect.top, 5, rect.height),
+        rect.width > 0
+            ? rrect
+            : RRect.fromRectAndRadius(Rect.fromLTWH(rect.left, rect.top, 5, rect.height), Radius.zero),
```
Comment (domesticmouse) [inline super_text_layout/lib/src/text_selection_layer.dart:179]: @matthew-carroll I'm unsure what should be done here. Should we round these corners as well?

Comment (matthew-carroll) [reply super_text_layout/lib/src/text_selection_layer.dart:179]: Can you provide some context? What's the code we're looking at, what does it do now, and what are you proposing that it should do?

Comment (domesticmouse) [reply super_text_layout/lib/src/text_selection_layer.dart:179]: The code I'm looking at is the edge case for when the `rect` has zero width. The code currently just creates a rectangle of width 5. I'm wondering if this edge case should round the edge case created rectangle the same as the other rects. There currently isn't a golden test for this as changing it from not rounding the edge case rect to rounding it doesn't break the test suite.

**Comment by angelosilvestre** [comment]
@domesticmouse I cannot push to this branch, so I created a PR to add the updated goldens. Could you please download the goldens and update this PR?

**Comment by domesticmouse** [comment]
@angelosilvestre I believe I have pulled the content from https://github.com/superlistapp/super_editor/pull/2136 into this PR.

File: super_text_layout/lib/src/text_selection_layer.dart (hunk 3 above)
Comment (matthew-carroll) [reply super_text_layout/lib/src/text_selection_layer.dart:179]: > The code I'm looking at is the edge case for when the rect has zero width

Is this the same thing as the optional policy that we have to display a selection when empty? If so, I thought I saw that you already rounded the empty selection rectangle?

If this is something different, can you let me know where/when this 5px wide rectangle currently appears?

**Comment by domesticmouse** [comment]
> @domesticmouse I cannot push to this branch, so I created a PR to add the updated goldens. Could you please download the goldens and update this PR?

Reading the code, this path is only triggered on empty lines, when this code expands the selection box from zero to five pixels wide.

**Comment by domesticmouse** [comment]
Please land. I'd like to use this for my presentation in India in a couple of weeks.

**Comment by matthew-carroll** [comment]
I haven't had a chance to look into that empty selection code path, but I'll merge as-is so you can use it.

**Review by matthew-carroll — APPROVED** [review]
LGTM

**Comment by matthew-carroll** [comment]
@angelosilvestre can you cherry pick this, and also close your other PR if it's no longer needed?
